e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -233,24 +214,95 @@ func (pf *PortForwarder) start() error {
-				}
-				// Watch automatically retries, so this shouldn't happen unless
-				// there's a terminal error.
-				pf.logger.Infof("Watch exited, restarting port-forward for %s/%s: %v", pf.ns, pf.deployment, err)
-			}
-			if err := pf.portForwardToDeployment(true); err != nil {
-				pf.logger.Info(err)
-				// this is a potentially fatal error since we failed to restart the port
-				// forward. we log the error and return in case the test is already done
-				// with the port forward. the test will otherwise error the next time it
-				// tries to use the local port.
+			if err := pf.waitAndRestart(); err != nil {
+				pf.logger.Infof("PortForwarder finished: %v", err)
 				return
 			}
 		}
 	}()
 	return nil
 }
+
+// waitAndRestart will block until seeing an indicator that the port-forward is
+// no longer healthy. If that occurs, it will restart the port-forward
+func (pf *PortForwarder) waitAndRestart() error {
+	err := pf.waitForFailure()
+	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+		return fmt.Errorf("exiting - PortForwarder for pod %s/%s received done signal", pf.ns, pf.pod)
+	}
+	pf.logger.Infof("received signal for port-forward for %s/%s. err: %v", pf.ns, pf.pod, err)
+	if err := pf.portForwardToDeployment(true); err != nil {
+		// this is a potentially fatal error since we failed to restart the port
+		// forward. we log the error and return in case the test is already done
+		// with the port forward. the test will otherwise error the next time it
+		// tries to use the local port.
+		return err
+	}
+	return nil
+}
+
+// waitForFailure waits for any failure condition which indicates the port-forward
+// must be restarted. The port-forward must be restarted if:
+// - the subprocess is killed
+// - the pod is restarted
+func (pf *PortForwarder) waitForFailure() error {
+	ctx, cancel := context.WithCancel(pf.ctx)
+	cmdChan := functionWithChannel(func() error {
+		// cmd.Wait does not accept a context, so create a wrapper that cancels with the ctx
+		cmdWaitChan := functionWithChannel(pf.cmd.Wait)
+		var chanErr error
+		select {
+		case <-ctx.Done():
+			chanErr = ctx.Err()
+		case err := <-cmdWaitChan:
+			chanErr = err
+		}
+		// kill the subprocess. This should force the cmd.Wait goroutine to return, if it hasn't already.
+		if err := pf.cmd.Process.Kill(); err != nil && !errors.Is(err, os.ErrProcessDone) {
+			// potentially bad if this fails, cmd.Wait may never return?
+			pf.logger.Infof("failed to kill port forward process", err)
+		} else if err == nil {
+			pf.logger.Infof("stopped port-forward process for %s/%s", pf.ns, pf.pod)
+		}
+		// wait for cmdWaitChan to close, i.e. cmd.Wait returns
+		<-cmdWaitChan
+		return chanErr
+	})
+	podChan := functionWithChannel(func() error {
+		err := pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns,
+			testwatcher.WatchTimeout(0), testwatcher.WatchContext(ctx))
+		// We're using WatchForNotFound to detect if the Pod is deleted, but a nil
```

Comment:
nit: this comment is mildly confusing. What we need is a function description. So maybe just move it up and say "Wait for the Pod to be deleted, then return an error to trigger port-forward restart"

Reply:
done